From 614406dc2f321973091be75d7f9d4763c8839fae Mon Sep 17 00:00:00 2001
From: Sam Elliott <aelliott@qti.qualcomm.com>
Date: Mon, 6 Jul 2026 11:13:11 -0700
Subject: [PATCH] [riscv][crt] Add Support for Xqccmt

Xqccmt is a vendor-specific variant of Zcmt with additional
capabilities. Enabling it in the CRT is exactly the same as for Zcmt -
the `jvt` csr nees to be initialised with `__jvt_base$`.

Signed-off-by: Sam Elliott <aelliott@qti.qualcomm.com>
(cherry picked from commit 3faeb852f82ea88395847b33f4680d43ac96a022)
---
 picocrt/machine/riscv/crt0.c | 8 ++++----
 test/test-riscv-jvt.c        | 4 ++--
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/picocrt/machine/riscv/crt0.c b/picocrt/machine/riscv/crt0.c
index fc75d7cb3..a194681aa 100644
--- a/picocrt/machine/riscv/crt0.c
+++ b/picocrt/machine/riscv/crt0.c
@@ -170,7 +170,7 @@ _trap(void)
 #endif
             ".option	pop");
 
-#ifdef __riscv_zcmt
+#if defined(__riscv_zcmt) || defined(__riscv_xqccmt)
     __asm__(".weak __jvt_base$\n"
 #ifdef __riscv_cmodel_large
             "ld t0, .trap_jvt_base\n"
@@ -202,7 +202,7 @@ _trap(void)
     __asm__(".align 3\n"
             ".trap_sp: .dword __stack\n" /* FIXME: this is __heap_end in non-large code models. */
             ".trap_gp: .dword __global_pointer$\n"
-#ifdef __riscv_zcmt
+#if defined(__riscv_zcmt) || defined(__riscv_xqccmt)
             ".trap_jvt_base: .dword __jvt_base$\n"
 #endif
     );
@@ -252,7 +252,7 @@ _start(void)
                 "csrwi	vxrm, 1");
 #endif
 
-#ifdef __riscv_zcmt
+#if defined(__riscv_zcmt) || defined(__riscv_xqccmt)
     __asm__(".weak __jvt_base$\n"
 #ifdef __riscv_cmodel_large
             "ld t0, .start_jvt_base\n"
@@ -287,7 +287,7 @@ _start(void)
 #ifdef CRT0_SEMIHOST
             ".start_trap: .dword _trap\n"
 #endif
-#ifdef __riscv_zcmt
+#if defined(__riscv_zcmt) || defined(__riscv_xqccmt)
             ".start_jvt_base: .dword __jvt_base$\n"
 #endif
             );
diff --git a/test/test-riscv-jvt.c b/test/test-riscv-jvt.c
index e9539bd08..cb22a7c22 100644
--- a/test/test-riscv-jvt.c
+++ b/test/test-riscv-jvt.c
@@ -36,7 +36,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 
-#if !defined(__riscv) || !defined(__riscv_jvt)
+#if !defined(__riscv) || (!defined(__riscv_zcmt) && !defined(__riscv_xqccmt))
 
 const char    *get_message(void);
 
@@ -77,7 +77,7 @@ main(void)
     /*
      * This is a RISC-V specific feature.
      */
-#if !defined(__riscv) || !defined(__riscv_zcmt)
+#if !defined(__riscv) || (!defined(__riscv_zcmt) && !defined(__riscv_xqccmt))
     printf("jvt not supported for this target\n");
     return 77;
 #else
-- 
2.51.0

